### DISCUSSION TITLE
Why does an intermediate snapshot have to be destroyed to rollback an earlier snapshot

### DISCUSSION THREAD

**Discussion by unkownerror1** [openzfs/zfs#13667] [Q&A]
I use ZFS version 0.8.6.
I have 3 snapshots，xxx/fs2@1, xxx/fs2@2, xxx/fs2@3. when i rollback xxx/fs2@2， xxx/fs2@3 will be destroyed, why not keep  xxx/fs2@3

![image](https://user-images.githubusercontent.com/109565529/179654234-f29d27de-ad87-473c-b398-6cbbc01c9f61.png)

**Comment by IvanVolosyuk** [comment]
Currently all the snapshots are organized in a sequence. What you suggest
will make them organized in a tree: one parent snapshot can have multiple
children snapshots. It makes management more complicated. You can actually
keep the intermediate snapshot if you want by using 'zfs clone, zfs
promote'. This will explicitly instruct zfs that you would like to have a
tree of snapshots.

**Reply by unkownerror1** [reply]
I do not understand why following my suggestion would make them organized in a tree

**Comment by IvanVolosyuk** [comment]
If you keep @3 and rollback to @2 and then snapshot @4. This will make @3
and @4 children of @2.

**Reply by unkownerror1** [reply]
if clone  filesystem ，Snapshots cannot be destroyed. What I need is a snapshot that can be destroyed, do you have any better suggestions

**Reply by Finix1979** [reply]
As I know, the original snapshot could not be destroyed if there is clone based on it even you modify whole data of the clone.

**Reply by IvanVolosyuk** [reply]
Well, the snapshot which used by two different filesystems (main + clone) shouldn't be deleted because it effectively branch point. Once you cut all the branches from it you can delete it. With 'zfs promote' you can change the ownership of the shared sequence of snapshots, e.g. move ownership to a clone and delete original. This way you are not locked on keeping the snapshot forever if that what you mean.

**Comment by rincebrain** [comment]
Currently, AIUI, the way ZFS decides whether something can be actually deleted is "are there any snapshots present between [creation of thing] and [deletion of thing]", which relies on there being one timeline of things with one creation and one deletion time.

If, given snapshots A,B,C,D, you rolled back to snapshot B, and then made changes, you would then have timeline A,B,C,D, and also timeline A,B,E, and the requirement for lifetimes of things to only have a single start and stop is broken and significantly more complicated to reason about.

(Conceptually, one could use something like #13392 to implement a clone-like operation where you end up with the equivalent of `cp --reflink` on an entire dataset, or do the same magic trick but on the thing you were planning to roll back and destroy to make an alternate timeline...but the devil would be in the details.)
